bolt/lib/Rewrite/RewriteInstance.cpp

```diff
@@ -1276,6 +1282,68 @@ void RewriteInstance::discoverFileObjects() {
   registerFragments();
 }
 
+Error RewriteInstance::discoverRtFiniAddress() {
+  // If FiniFunctionAddress is already set, we got if from DT_FINI. We use
+  // DT_FINI instead of DT_FINI_ARRAY if it's available.
+  if (BC->FiniFunctionAddress)
+    return Error::success();
+
+  if (!BC->FiniArrayAddress || !BC->FiniArraySize) {
+    return createStringError(
+        std::errc::not_supported,
+        "Instrumentation needs either DT_FINI or DT_FINI_ARRAY");
+  }
+
+  if (*BC->FiniArraySize < BC->AsmInfo->getCodePointerSize()) {
+    return createStringError(std::errc::not_supported,
+                             "Need at least 1 DT_FINI_ARRAY slot");
+  }
+
+  ErrorOr<BinarySection &> FiniArraySection =
+      BC->getSectionForAddress(*BC->FiniArrayAddress);
+  if (auto EC = FiniArraySection.getError())
+    return errorCodeToError(EC);
+
+  BC->FiniArraySection = &*FiniArraySection;
+
+  if (const Relocation *Reloc = FiniArraySection->getDynamicRelocationAt(0)) {
+    BC->FiniFunctionAddress = Reloc->Addend;
+    return Error::success();
+  }
+
+  if (const Relocation *Reloc = FiniArraySection->getRelocationAt(0)) {
+    BC->FiniFunctionAddress = Reloc->Value;
+    return Error::success();
+  }
+
+  return createStringError(std::errc::not_supported,
+                           "No relocation for first DT_FINI_ARRAY slot");
+}
+
+void RewriteInstance::updateRtFiniReloc() {
+  if (!BC->FiniArraySection)
```

Comment:
To be honest I don't really like FiniArraySection variable. First of all it is a bit confusing, since we might have fini array section, but we don't set it since we found DT_FINI to use instead. Renaming it for RT purposes seems to be not the best solution too. Please consider adding optional pointer argument for discoverRtFiniAddress to get the FiniArraySection if needed or returning tuple for example, checking both return value and the arg != nullptr,  it seems to be more straightforward solution to me, what do you think?

Reply:
I agree that `FiniArraySection` is a bit confusing. The only reason I added it was to avoid having to call `getSectionForAddress` a second time in `updateRtFiniReloc`. I don't think this is worth it though so I removed `FiniArraySection` and simply called `getSectionForAddress` again. Does this address your concern?

Reply:
Not exactly, currently you at discoverRtFiniAddress you also set FiniFunctionAddress.  As I understand it is caused by the internal logic, which uses it to call the original function from new instrumentation fini.

I didn't notice it before to be honest, but the problem with your current change that after you set the FiniFunctionAddress you can't distinguish between the case where DT_FINI was originally in the binary and you don't have to patch DT_FINI_ARRAY and your case where you should. Correct me if I'm wrong. 

So you need to check that FiniFunctionAddress == Reloc->Value of the first dynamic slot to continue patching at least. 
My suggestion is to use FiniFunctionAddress for DT_FINI only (probably fix the comment about it) and add FiniFunctionRtAddress and check that FiniFunctionAddress != FiniFunctionRtAddress with the comment that DT_FINI is fixed during patching DYNAMIC. After it I would probably add assert to BC->FiniArrayAddress && BC->FiniArraySize just in case. And one more assert after successfully taking takeDynamicRelocationAt with BC->FiniFunctionRtAddress == Reloc->Addend; This way it seems to be less frustrating and also gives ability to rewrite comparing the addresses if would want to use not only DT_FINI and DT_FINI_ARRAY in the future (unlikely, but still).

Reply:
> I didn't notice it before to be honest, but the problem with your current change that after you set the FiniFunctionAddress you can't distinguish between the case where DT_FINI was originally in the binary and you don't have to patch DT_FINI_ARRAY and your case where you should. Correct me if I'm wrong.

Ugh, nothing to correct, you're right of course :) I completely missed this.

I fixed this similar to what you suggest by introducing a new `FiniAddress` variable. I think this makes things more consistent:
- `DT_FINI` -> `FiniAddress`;
- `DT_FINI_ARRAY` -> `FiniArrayAddress`;
- `DT_FINI_ARRAYSZ` -> `FiniArraySize`.

The original `FiniFunctionAddress` is kept as-is and is set either to `FiniAddress` (if it has a value) or to one of the relocs (if `FiniAddress` has no value). So the way to distinguish between using `DT_FINI` or `DT_FINI_ARRAY` is by checking if `FiniAddress` has a value.

Do you think this addresses the issue? In particular, I wondered about this:

> So you need to check that FiniFunctionAddress == Reloc->Value of the first dynamic slot to continue patching at least.

I believe this is not necessary because if `FiniAddress` does not have a value, the first reloc *must* be the one we want to patch (as this was checked in `discoverRtFiniAddress`).

Reply:
Thanks for the fix! It seems to be good now. Only the tests seems to left. Just a hint, I would build the single test with 3 different states: only DT_FINI, only DT_FINI_ARRAY, and both for pie and non-pie cases, each time checking the static and if needed dynamic relocs were updated.

>> I believe this is not necessary 

Well it is just extra check that everything is in consistent state, I don't really insist on it, but I would stay with it :)